From 1dd86c699bea406ba6f88d773cf88776818ea229 Mon Sep 17 00:00:00 2001
From: Florian Westphal <fw@strlen.de>
Date: Fri, 27 Mar 2020 03:24:49 +0100
Subject: [PATCH] netfilter: nf_queue: prefer nf_queue_entry_free

Instead of dropping refs+kfree, use the helper added in previous patch.

Signed-off-by: Florian Westphal <fw@strlen.de>
Signed-off-by: Pablo Neira Ayuso <pablo@netfilter.org>
(cherry picked from commit 28f715b9e6dd7cbf07c2aea913fea7c87a56a3b5)

[Quentin: conflicts due to the nf_queue_entry_extra shadow variable and
the extra nf_queue_entry_get_refs_bool() error path, both of which still
need to free the shadow variable without dropping refs: kept a goto err
label for them.]

Signed-off-by: Quentin Casasnovas <quentin.casasnovas@vates.tech>
---
 net/netfilter/nf_queue.c | 31 ++++++++++++-------------------
 1 file changed, 12 insertions(+), 19 deletions(-)

diff --git a/net/netfilter/nf_queue.c b/net/netfilter/nf_queue.c
index fdbf5b43c100..dc616bc6367f 100644
--- a/net/netfilter/nf_queue.c
+++ b/net/netfilter/nf_queue.c
@@ -212,19 +212,17 @@ static int __nf_queue(struct sk_buff *skb, const struct nf_hook_state *state,
 		      const struct nf_hook_entries *entries,
 		      unsigned int index, unsigned int queuenum)
 {
-	int status = -ENOENT;
-	struct nf_queue_entry_extra *extra = NULL;
+	struct nf_queue_entry_extra *extra;
 	struct nf_queue_entry *entry = NULL;
 	const struct nf_queue_handler *qh;
 	struct net *net = state->net;
 	unsigned int route_key_size;
+	int status;
 
 	/* QUEUE == DROP if no one is waiting, to be safe. */
 	qh = rcu_dereference(net->nf.queue_handler);
-	if (!qh) {
-		status = -ESRCH;
-		goto err;
-	}
+	if (!qh)
+		return -ESRCH;
 
 	switch (state->pf) {
 	case AF_INET:
@@ -239,15 +237,13 @@ static int __nf_queue(struct sk_buff *skb, const struct nf_hook_state *state,
 	}
 
 	entry = kmalloc(sizeof(*entry) + route_key_size, GFP_ATOMIC);
-	if (!entry) {
-		status = -ENOMEM;
-		goto err;
-	}
+	if (!entry)
+		return -ENOMEM;
 
 	extra = shadow_var_alloc(entry, NF_QUEUE_ENTRY_EXTRA, sizeof(*extra), GFP_ATOMIC);
-	if (extra == NULL) {
-		status = -ENOMEM;
-		goto err;
+	if (!extra) {
+		kfree(entry);
+		return -ENOMEM;
 	}
 	*extra = (struct nf_queue_entry_extra) {
 		.skb_dev = skb->dev,
@@ -282,19 +278,16 @@ static int __nf_queue(struct sk_buff *skb, const struct nf_hook_state *state,
 	}
 
 	status = qh->outfn(entry, queuenum);
-
 	if (status < 0) {
-		nf_queue_entry_release_refs(entry);
-		goto err;
+		nf_queue_entry_free(entry);
+		return status;
 	}
 
 	return 0;
 
 err:
-	if (extra)
-		shadow_var_free(entry, NF_QUEUE_ENTRY_EXTRA);
+	shadow_var_free(entry, NF_QUEUE_ENTRY_EXTRA);
 	kfree(entry);
-
 	return status;
 }
 
